### PR TITLE
[feature/mailer-optimization] Add interfaces and Events

### PR DESCRIPTION
**Pull request by lx-wnk**
### 1. Why is this change necessary?
To catch and enhance mails 

### 2. What does this change do, exactly?
- Add events to MailService.php
- Add Interfaces for decoration

### 3. Describe each step to reproduce the issue or behaviour.
///

### 4. Please link to the relevant issues (if any).
///

### 5. Checklist

- [ ] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [ ] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### CODE REVIEW

**Comment by hlohaus** [comment]
Nice. But you can catch mails:
https://github.com/hlohaus/SheMailer/blob/master/src/Resources/config/services.xml#L8
:)
A question: Can i access then custom fields?
https://github.com/shopware/platform/pull/340

**Comment by lx-wnk** [comment]
@hlohaus
That might be possible but if you want to handle it config based it would be easier to use the events. I think also that more events in that case would provide a better usability for enhancements.

I think you should be able to but honestly I have not tried it yet. Although maybe the events needs to be enhanced

**Comment by hlohaus** [comment]
Hi, thanks for you answer and pull request.
if you hadn't made it, I would have made it.
Can you integrate my pull request in yours?
Then it possible to something like that: 

https://github.com/hlohaus/platform/commit/20285089150daa3dfdbda0f29da1510d6a4e76b2#diff-98867856d544c8cd22262d2d0b47489bR169

@shopwareBot What do you think?

**Comment by lx-wnk** [comment]
@hlohaus you're welcome.
I'll integrate the PR as soon as it is tagged for or just merged - whatever happens first.

**Comment by ssltg** [comment]
Hi Alex,
thank you for your contribution. We imported your PR and will merge it soon.
👍

File: src/Core/Content/MailTemplate/Service/Event/MailBeforeSendEvent.php [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+    public function getTemplateData(): array
+    {
+        return $this->templateData;
+    }
+
+    public function getLogData(): array
+    {
+        return [
+            'subject' => $this->subject,
```
Comment (hlohaus) [inline src/Core/Content/MailTemplate/Service/Event/MailBeforeSendEvent.php]: maybe wrong?

Comment (lx-wnk) [reply src/Core/Content/MailTemplate/Service/Event/MailBeforeSendEvent.php]: I made some changes, could you take a look again?

File: src/Core/Content/MailTemplate/Service/MailerFactoryInterface.php [hunk 2]
```diff
@@ -0,0 +1,8 @@
+<?php declare(strict_types=1);
+
+namespace Shopware\Core\Content\MailTemplate\Service;
+
+interface MailerFactoryInterface
+{
+    public function create(\Swift_Message $message): \Swift_Mailer;
```
Comment (hlohaus) [inline src/Core/Content/MailTemplate/Service/MailerFactoryInterface.php]: maybe wrong

**Review by hlohaus — CHANGES_REQUESTED** [review]
additionally customFields can be integrated.

**Review by hlohaus — APPROVED** [review]

**Comment by lx-wnk** [comment]
> additionally customFields can be integrated.

I think the current implementation will work for this too. If you want to enhance the message or parts of it you could just subscribe to the event and handle the custom fields there. If it is not explicit needed I think an implementation would be unnecessary.

**Comment by hlohaus** [comment]
I tried it out. There was one bug and it was not possible to get the custom fields or the mail template.
But this works:
https://github.com/hlohaus/platform/commit/f8ca94c1be436f7b4fe9e3be4635a5c02e21e5bc

File: src/Core/Content/MailTemplate/Service/MailService.php [hunk 3]
```diff
@@ -74,8 +76,8 @@ class MailService
     public function __construct(
```
Comment (hlohaus) [inline src/Core/Content/MailTemplate/Service/MailService.php:76]: implements MailServiceInterface...

**Review by hlohaus — CHANGES_REQUESTED** [review]

File: src/Core/Content/MailTemplate/Service/MailService.php (hunk 3 above)
Comment (lx-wnk) [reply src/Core/Content/MailTemplate/Service/MailService.php:76]: Is added

**Review by hlohaus — APPROVED** [review]

**Comment by lx-wnk** [comment]
@hlohaus and @ssltg could you both please review my changes again?
I changed the target to 6.1 and added thre template id and the custom fields to the databag according to a discussion with @hlohaus

**Review by hlohaus — APPROVED** [review]

**Comment by mitelg** [comment]
thanks for your contribution @jinnoflife :+1: :tada: :blue_heart:

**Comment by hlohaus** [comment]
@shopwareBot please import https://github.com/shopware/platform/commit/af9f23bcb3ddde7c11a8849541842253bd4838b0
not 
https://github.com/shopware/core/commit/cd215853c5352f1193398aba75100b56814e6d25.
the second are broken.

**Comment by mitelg** [comment]
hey @hlohaus 
we have reworked the PR a bit anyway. Would you mind having a look at the changes and see if they still fit your needs? Otherwise another PR as follow up would be nice :blush: :+1:

**Comment by lx-wnk** [comment]
Hey @mitelg is there something missing or to do for @hlohaus and/or me?

I saw that the [6.2 RC is without](https://github.com/shopware/platform/blob/master/CHANGELOG-6.2.md) these changes.

**Edit:**
Never mind. Just saw that the 6.2-tag contains everything 🙂

**Pull request closed without merging**